overlord/servicestate/servicemgr.go

```diff
@@ -42,8 +53,215 @@ func Manager(st *state.State, runner *state.TaskRunner) *ServiceManager {
+		// about (LP #1924805) here was never triggered
+		return nil
+	}
+
+	candidateAppsToRestartBySnap := make(map[*snap.Info][]*snap.AppInfo)
+
+	for sn, apps := range modified {
+		for _, app := range apps {
+			// get the InactiveEnterTimestamp for the service
+			t, err := sysd.InactiveEnterTimestamp(app.ServiceName())
+			if err != nil {
+				return err
+			}
+
+			// check if this unit entered the inactive state between the time
+			// range, but be careful about time precision here, we want an
+			// inclusive range i.e. [lower,upper] not (lower,upper) in case the
+			// time that systemd saves these events as is imprecise or slow and
+			// things get saved as having happened at the exact same time
+			if !t.Before(lowerTimeBound) && !t.After(upperTimeBound) {
+				candidateAppsToRestartBySnap[sn] = append(candidateAppsToRestartBySnap[sn], app)
+			}
+		}
+	}
+
+	// Second loop actually restarts the services per-snap by sorting them and
+	// removing disabled services. Note that we could have disabled services
+	// here because a service could have been running, but disabled when snapd
+	// was refreshed, hence it got killed, but we don't want to restart it,
+	// since it is disabled, and so that disabled running service is just SOL.
+	for sn, apps := range candidateAppsToRestartBySnap {
+		// TODO: should we try to start as many services as possible here before
+		// giving up given the severity of the bug?
+		disabledSvcs, err := wrappers.QueryDisabledServices(sn, progress.Null)
+		if err != nil {
+			return err
+		}
+
+		startupOrdered, err := snap.SortServices(apps)
+		if err != nil {
+			return err
+		}
+
+		// TODO: what to do about timings here?
+		nullPerfTimings := &timings.Timings{}
+		if err := wrappers.StartServices(startupOrdered, disabledSvcs, nil, progress.Null, nullPerfTimings); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (m *ServiceManager) ensureSnapServicesUpdated() (err error) {
+	m.state.Lock()
+	defer m.state.Unlock()
+	if m.ensuredSnapSvcs {
+		return nil
+	}
+
+	// only run after we are seeded
+	var seeded bool
+	err = m.state.Get("seeded", &seeded)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	if !seeded {
+		return nil
+	}
+
+	// we are seeded, now we need to find all snap services and re-generate
+	// services as necessary
+
+	// ensure all snap services are updated
+	allStates, err := snapstate.All(m.state)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+
+	// if we have no snaps we can exit early
+	if len(allStates) == 0 {
+		m.ensuredSnapSvcs = true
+		return nil
+	}
+
+	snapsMap := map[*snap.Info]*wrappers.SnapServiceOptions{}
+
+	for _, snapSt := range allStates {
+		info, err := snapSt.CurrentInfo()
+		if err != nil {
+			return err
+		}
+
+		// don't use EnsureSnapServices with the snapd snap
+		if info.Type() == snap.TypeSnapd {
+			continue
+		}
+
+		// TODO: handle vitality rank here too from configcore
```

Comment:
this probably needs a prereq, as it involves the fact that only snapstate knows how to compute this atm :/

Reply:
I proposed #10171 for this